### ISSUE TITLE
Add a tutorial for new users

### ISSUE THREAD

**Issue by ipstone** [gokcehan/lf#94]
Hello, 

lf is really awesome! - Can we add a short tutorial for new users? such as

1. Installation - grab it

2. Config - copy example config to ~/.config/lf/lfrc `mkdir -p ~/.config/lf/ && cp lfrc.example ~/.config/lf/lfrc`
    The example lfrc. example is already very nice, it would be nice to add some more commonly used mappings, so that it will be easier/as well as instrumental.

3. Some directly available functions with the sample lfrc:

- move up/down in current folder: j/k
- move to upper level/lower level of folder: h/l
- Show/hide hidden files: zh
- run shell commend: <enter>
- select files/toggle: <space>

3. If someone can do a video and linked in the tutorial it would be nice. 'lf' is much faster than ranger for me on linux servers. In depth demonstration of using client/server etc. would be very instrumental.

Thanks,
Isaac

**Comment by gokcehan** [comment]
@ipstone Tutorial sounds nice. Maybe someone can create a wiki page and we can link it from the readme. For video demonstration, I think we can also consider embedded animated gifs with annotated keys. Maybe a collection of short gifs (2-5 secs) for those common operations you mention would be nice.

**Comment by ipstone** [comment]
All sounds great. I actually tried to create a wiki page, but noticed that our wiki here is not editable.  

I am not sure if github wiki can also use pull request - if it does, then I can make a copy of the repo, and edit wiki there for you to pull in?

Using gifs is also a great idea - though I am not sure if it's easier than video demonstration. Will think more about it. thanks

**Comment by TeddyDD** [comment]
Publicly editable wiki would be nice for sharing tips and tricks.

**Comment by TeddyDD** [comment]
Publicly editable wiki would be nice for sharing tips and tricks.

**Comment by TeddyDD** [comment]
Publicly editable wiki would be nice for sharing tips and tricks.

**Comment by TeddyDD** [comment]
Publicly editable wiki would be nice for sharing tips and tricks.

**Comment by gokcehan** [comment]
@TeddyDD @ipstone Sorry I wasn't aware there is such a setting. I think I have now changed wiki to be public. You should be able to create or edit pages. Tell me if the problem persists.

@ipstone For videos, you can use whatever you like. I'm just brainstorming. By the way, I think I had seen a ranger tutorial in youtube a while back ago which you may find interesting.

**Comment by gokcehan** [comment]
Also maybe I should mention a few things to give some idea.

Some people like to use asciinema for screencasts:

https://asciinema.org/

This seems nice but last time I checked for another project of mine I couldn't find a way to display pushed keys with this. So I settled on using animated gifs recorded using [peek](https://github.com/phw/peek) with [screenkey](https://github.com/wavexx/screenkey) to show keys. I got the idea from screencasts in kakoune website:

http://kakoune.org/

Kakoune also has a cool builtin ncurses asistant feature to show pushed keys so they can also use asciinema easily without an issue:

https://delapouite.github.io/kakoune-tv/

**Comment by ipstone** [comment]
@gokcehan @TeddyDD Thank you! The tutorial page was just created with some simple content - I will add more new user tips once get more experience with lf. 

We may consider closing this ticket item.

**Comment by gokcehan** [comment]
@ipstone Thanks for the page. I think we can keep this issue open for a while so it can attract potential contributors. Once we have more content on the page we can link it from the readme and maybe close this issue.

**Comment by ipstone** [comment]
@gokcehan Sounds great

**Comment by ipstone** [comment]
@gokcehan One quick question for windows: where should the lfrc file reside? thanks

**Comment by ipstone** [comment]
Thanks @gokcehan - perhaps we can add a new feature to prompt the user to have a 'default' settings when lf is first run (such as if there's no lfrc file exists)? 

It will make lf much easier for new users. Honestly, recently I began to use ranger more (on a fast machine), and quite surprised how the default setting can do so much (quite comprehensive info/help/keybindings for new users). lf is much faster and efficient, if new user can get a comprehensive default setting, it would help adoption.

**Comment by gokcehan** [comment]
@ipstone I was thinking of the same thing. The other day I copied my binary to a remote machine and tried to use it without a config file and it was barely any usable. We should have much stronger defaults such that maybe 90% of the use cases should be possible without a config file. I was thinking of setting some of the things in the example config file as default. Advanced users can always unmap or override these defaults when they want. What I have in mind is bindings for toggling options and info, different sorting types, common directory shortcuts, editor and pager bindings using `$EDITOR` and `$PAGER` variables with fallbacks to `vi` and `less`, maybe using a default file opener such as `start` on windows, `open` on mac, and `xdg-open` on others. I will try to add some of these things as soon as possible.

**Comment by ipstone** [comment]
@gokcehan Your plan sounds excellent, this is awesome! 
An easy alternative fix would be to just check if the config exist, if not download it from 'github' and save in the user's config location (or build in the config file in the lf binary, and save it to disk). 

A related needed example is the example of using lf with screen/tmux. The server/client is an excellent design, but new user may have trouble understand how to use it. A simple example of using lf with screen or tmux would help to clarify. If you have a quick example, we can put it on the wiki. Thanks

**Comment by TeddyDD** [comment]
> build in the config file in the lf binary, and save it to disk
Personally I think that would be great solution. It does not require a lot of code and it does not force users to unmap existing keybindings if they want to start from scratch.

**Comment by gokcehan** [comment]
Sorry for late reply.

I like the sound of this idea though I'm not so sure about automatically creating a configuration file. I feel like it should be possible to operate without a configuration file at all. A midway solution would be to only assign default keybindings if a configuration file is not present, kind of like default configuration script introduced in vim8.

Embedding files in the binary is a hack that I'm not very proud of. Honestly, I was hoping to get rid of it for the documentation at some point though I don't have a solid replacement strategy in my mind yet. Maybe we can automatically generate current settings and keybindings and leave the rest as an online documentation. In any case, our documentation can use a lift up but that's another issue.

For this idea, maybe we should also discuss whether we shall leave the rest of the default keybindings as it is or leave it to the user to assign them. The way it currently works is to have a default keybinding for most builtin commands and leave the rest alone, but this is kind of arbitrary from a user perspective. If we remove all default keybindings then we rely too much on the user to have an appropriate keybinding configuration (i.e. if you forgot to assign `read` and `quit` commands, you can't even quit the program). Even then, it will still be opinionated a little bit since we still have default option values.

In the end, two things are important. First, we should have strong defaults such that a user should be able to do almost everything without changing anything. It is best to keep in mind that there will always be some users which won't be satisfied with all of the default settings no matter what we choose. So second, it should be possible to reconfigure everything with a configuration file when desired. The good thing is, currently it is possible to override a keybinding without unmapping the default keybinding first except when there is a common subset. There is an open issue for the exceptional case which we may prioritize if needed.

@ipstone @TeddyDD Are you worried, these additional keybindings we are discussing are too much opinionated? Will these make it too much of a hassle to configure to your own desire? I still feel like we can settle on some defaults that should work more or less for all no matter how arbitrary it feels.

**Comment by TeddyDD** [comment]
> The good thing is, currently it is possible to override a keybinding without unmapping the default keybinding first 

That's great, calling unmap for everything would be annoying :) 

I like that lf is pretty bare-bones by default. I know what every key do since wrote config myself (except default keys but they are well documented). 
When I open ragner I feel like walking on minefield with all that preconfigured keybindings. Part of the problem with ranger is that help system is not great and removing all keybindings is a lot of hassle.

> Are you worried, these additional keybindings we are discussing are too much opinionated?

Common directory shortcuts are little too much (damn, I hate programs that assume I keep downloaded files in ~/Downloads...) , everything else sounds reasonable.

To sum up: I think default config would be great as long as it's well documented and easy to change.

**Comment by gokcehan** [comment]
@ipstone I think `:help` might be a good idea. To add what you are saying, currently there is no way to check the current values of options, so maybe it would be nice if we can generate the help page with the current values of options.

**Comment by gokcehan** [comment]
I have now added screencasts to the tutorial wiki and cleaned it up a little bit. Closing this issue now.
